### DISCUSSION TITLE
"Connect-only" WebSocket use case with two-threaded bidirectionnal communication

### DISCUSSION THREAD

**Discussion by MatouCode** [curl/curl#12660] [Q&A]
Hello, 

I have a high level WebSocket client with two use cases, quite similar to what libcurl proposes : 

- A "free" mode where the client performs the HTTP upgrade synchronously on one thread, and then the user may use the functions
_SendWebSocket_ and _DownloadWebSocket_ (a blocking call) freely.

- A "callback" mode where the user implements a few callbacks, similar to what the [JSR 356](https://jcp.org/aboutJava/communityprocess/mrel/jsr356/index.html) offers : OnOpen(), OnMessage(), OnClose(). The callbacks 
are executed on a separate thread. 

For now I played around with WinHTTP's WebSocket API (which actually also exist on Wininet, though not documented, just replace _WinHttpWebSocket_* by _InternetWebSocket_*.)

In my backend, the "send" operations are done with WinHttpWebSocketSend on the thread where the user calls my Send function, and the receive (WinHttpWebSocketReceive) is called in a loop on a download thread, which pushes data in a shared queue. This queue is read either by DownloadWebSocket, i.e. directly by the user, either by my callback thread before executing OnMessage(). I use WinHttp's asynchronous mechanisms to wait on the internet handle while nothing is happening on the receive side.

Now I want to replace WinHTTP by libcurl (at least when it won't be flagged as experimental), for portability and reliability reasons, and I'd like my use cases to be preserved. (For the waiting part, I guess a _curl_ws_poll_ would be useful, else I think I'd be getting into platform-dependent considerations if I wait/select/poll directly on the socket... but I know this is being considered.)

Hence my question : **is it possible to call curl_ws_send and curl_ws_recv respectively on two different threads, simultaneously?** This would make sense in view of the bidirectional philosophy of WebSockets - and because if I'm using WinHttp, or a pure socket, those send and recv are threadsafe, cf (https://learn.microsoft.com/en-us/windows/win32/winhttp/concurrency-in-winhttp#web-socket-send-and-receive-operations). But nothing actually confirms it in the doc, if I'm right.

Thanks in advance for your insights,

**Comment by dfandrich** [comment]
The libcurl APi does not allow simultaneous access to any given handle from
more than one thread. If you want to do something like that, you'd have to
ensure that each thread locks all access to libcurl calls to prevent *actual*
simultaneous access, and if you read other bugs here, you'll see that it's easy
to miss one which will cause bad behaviour.

**Reply by MatouCode** [reply]
This sounds like a regression to other WebSocket libraries like Wininet.

Bidirectionality is at the heart of WebSockets and perhaps libcurl should consider allowing parallel recv/sends. If I have no prior idea of what ​​the communication will look like, and if I put curl_ws_recv and curl_ws_send in mutual exclusion, I easily find myself blocking the send channel when I receive data and vice versa. Even though if I "steal" the socket after the upgrade, I could use recv and send in parallel. 

Maybe my feedback is naive, but as it is, the API doesn't seem natural or very practical to use.
